packages/network-controller/tests/NetworkController.test.ts

```diff
@@ -3535,6 +3535,352 @@ describe('NetworkController', () => {
+              },
+              async ({ controller }) => {
+                const fakeInfuraProvider = buildFakeInfuraProvider();
+                createInfuraProviderMock.mockReturnValue(fakeInfuraProvider);
+                const fakeInfuraSubprovider = buildFakeInfuraSubprovider();
+                SubproviderMock.mockReturnValue(fakeInfuraSubprovider);
+                const fakeMetamaskProvider = buildFakeMetamaskProvider([
+                  {
+                    request: {
+                      method: 'eth_chainId',
+                    },
+                    response: {
+                      result: '0x1337',
+                    },
+                  },
+                ]);
+                createMetamaskProviderMock.mockReturnValue(
+                  fakeMetamaskProvider,
+                );
+
+                controller.resetConnection();
+
+                const { provider } = controller.getProviderAndBlockTracker();
+                const promisifiedSendAsync = promisify(provider.sendAsync).bind(
+                  provider,
+                );
+                const { result: chainIdResult } = await promisifiedSendAsync({
+                  method: 'eth_chainId',
+                });
+                expect(chainIdResult).toBe('0x1337');
+              },
+            );
+          });
+
+          it('replaces the provider object underlying the provider proxy without creating a new instance of the proxy itself', async () => {
+            await withController(
+              {
+                state: {
+                  providerConfig: {
+                    type: networkType,
+                    // NOTE: This doesn't need to match the logical chain ID of
+                    // the network selected, it just needs to exist
+                    chainId: '0x9999999',
+                  },
+                },
+              },
+              async ({ controller }) => {
+                const fakeInfuraProvider = buildFakeInfuraProvider();
+                createInfuraProviderMock.mockReturnValue(fakeInfuraProvider);
+                const fakeInfuraSubprovider = buildFakeInfuraSubprovider();
+                SubproviderMock.mockReturnValue(fakeInfuraSubprovider);
+                const fakeMetamaskProvider = buildFakeMetamaskProvider();
+                createMetamaskProviderMock.mockReturnValue(
+                  fakeMetamaskProvider,
+                );
+
+                controller.providerConfig = {
+                  type: NetworkType.mainnet,
+                  chainId: '0x9999999',
+                };
+
+                const { provider: providerBefore } =
+                  controller.getProviderAndBlockTracker();
+                controller.resetConnection();
+                const { provider: providerAfter } =
+                  controller.getProviderAndBlockTracker();
+
+                expect(providerBefore).toBe(providerAfter);
+              },
+            );
+          });
```

Comment:
If you look at the tests for, say, `setProviderType`, there's a test `ensures that the existing provider is stopped while replacing it`. I think it'd be good to bring this test over as well since that behavior is also exercised when `resetConnection` is called.

I think we would also want to copy the `when an "error" event occurs on the new provider` describe block from `setProviderType`, and adapt them for `resetConnection`, since that behavior also applies here too.